File: behavioral/memento/document_editor/typescript/document.ts
// behavioral/memento/document_editor/typescript/document.ts
import { IMemento, ConcreteMemento } from './memento';

// The Originator holds some important state that may change over time.
export class Document {
    private content: string;

    constructor(initialContent: string = "") {
        this.content = initialContent;
        console.log(`Document initialized with: '${this.content}'`);
    }

    // Appends text to the document content.
    public write(text: string): void {
        console.log(`Writing: '${text}'`);
        this.content += text;
        console.log(`Current content: '${this.content}'`);
    }

    // Sets the document content directly (used for restoring).
    public setContent(content: string): void {
        this.content = content;
        console.log(`Content set to: '${this.content}'`);
    }

    // Returns the current content.
    public getContent(): string {
        return this.content;
    }

    // Saves the current state inside a memento.
    public save(): IMemento {
        const memento = new ConcreteMemento(this.content);
        console.log(`Saving state: '${this.content}'`);
        return memento;
    }

    // Restores the Originator's state from a memento object.
    public restore(memento: IMemento): void {
        this.content = memento.getState();
        console.log(`Restoring state to: '${this.content}'`);
    }
}

File: behavioral/memento/document_editor/python/document.py
from memento import Memento

class Document:
    """
    The Originator holds some important state that may change over time.
    It also defines a method for saving the state inside a memento and
    another method for restoring the state from it.
    """
    def __init__(self, initial_content: str = ""):
        self._content = initial_content
        print(f"Document initialized with: '{self._content}'")

    def write(self, text: str):
        """Appends text to the document content."""
        print(f"Writing: '{text}'")
        self._content += text
        print(f"Current content: '{self._content}'")

    def set_content(self, content: str):
        """Sets the document content directly (used for restoring)."""
        self._content = content
        print(f"Content set to: '{self._content}'")

    def get_content(self) -> str:
        """Returns the current content."""
        return self._content

    def save(self) -> Memento:
        """Saves the current state inside a memento."""
        memento = Memento(self._content)
        print(f"Saving state: '{self._content}'")
        return memento

    def restore(self, memento: Memento):
        """Restores the Originator's state from a memento object."""
        self._content = memento.get_state()
        print(f"Restoring state to: '{self._content}'")
